### PR TITLE
 [INTERNAL][I] Introduce SelectedEditorState

### PR DESCRIPTION
**Pull request by tobous**
Introduces the class SelectedEditorState which can be used to capture
and re-apply which editors are selected in the local IDE. Without such a
logic, opening a new editor as part of the internal Saros logic would
change the currently selected editor, thereby possibly hindering the
user.

Use the logic provided by SelectedEditorState instead of implementing
the logic directly in EditorManager.addProjectResources(IProject).

### CODE REVIEW

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        .getLogger(SelectedEditorState.class);
+
+    private final List<VirtualFile> selectedEditors;
+
+    private boolean hasCapturedState;
+
+    @Inject
+    private ProjectAPI projectAPI;
+
+    @Inject
```
[the remark this answers is not in the document]
Comment (tobous) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: I implemented this class to use it in the upcoming rework of [SharedResourcesManager](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/SharedResourcesManager.java) and [FileSystemChangeListener](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java). In both classes, we would also be missing Project (and, in the upcoming change, SharedResourcesManager would be missing ProjectAPI if it didn't have to hand it to FileSystemChangeListener).

My thinking was that I wanted to keep the usage of this class as simple as possible as it is meant as a general purpose utility class. I didn't want to pass these objects through the CTOR as it too often leads to a situation where other classes wanting to use this class would themselves have to use CTOR injection and add the needed objects as instance variables just to use this class.

So in short, I did not know which objects were present anyway in the classes using this class. I could just go by the current use cases, but then passing some the needed objects and using annotation injection for others seems like a weird halve-measure to me.

I see the point that this is just wasted computing time if all the needed objects are already present, but I still think this is a cleaner solution. But if you strongly disagree, I can change it.

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            VirtualFile newFile) {
+
+        int index = selectedEditors.indexOf(oldFile);
+
+        if (index != -1) {
+            selectedEditors.set(index, newFile);
+
+        } else {
+            log.warn("Could not replace " + oldFile + " with " + newFile
+                + " as the captured state does not contain the given file to "
```
[the remark this answers is not in the document]
Comment (tobous) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: Will add a check for `!hasCapturedState` and log a warning like in `applyCapturedState()`.

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (srossbach) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: The boilerplated solution would be to add a factory that has only one method. I really dislike the injection part as the PicoContainer library is discontinued since ages and there are potential deadlock pits especially when calling start and stop on such a container (as we do this not in the Application thread). This is also the reason why ISarosSession#getComponent() returns null when the session is in the stopping phase, simply to avoid deadlocking.

[the remark this answers is not in the document]
Comment (srossbach) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: Just forgot what I said, this will not help as this will result in a cyclic dependency.

Factory -> A, B, C

A -> Factory

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            return;
+        }
+
+        int index = selectedEditors.indexOf(oldFile);
+
+        if (index != -1) {
+            selectedEditors.set(index, newFile);
+
+        } else {
+            log.debug("Could not replace " + oldFile + " with " + newFile
```
Comment (tobous) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: I changed this log level to debug as there is currently no good logic to check whether a given file is selected (without involving a lot of IntelliJ logic). This means that this method will always be called, independent of whether the file to replace was actually selected to begin with.

I decided to do it this way around as the initial check whether the file is selected would probably take up more computing time than unnecessarily calling this method.

The only somewhat ugly remain is that this message will be logged in such cases, making it harder to determine when this case is actually a problem (caused by an inconsistency of some kind). But this is acceptable in my opinion.

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tobous) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: I am not sure what your current stance on the implementation using annotation injection is.
Is the lifecycle stuff really such a big problem, even though this is using the plugin container/context (instead of the much more short lived session context)?
If you would still like to avoid annotation injection, I will change it.

On a related note: How pressing is the whole PicoContainer being legacy/unmaintained thing in your opinion?

There are some alternatives for dependency injection libraries, but the switch will probably be very work intensive. Especially the regression testing to check that the switch didn't break anything. But, depending on the importance, this could be something we could be working on after the gradle stuff and the subsequent JDK upgrade is done.

(In general I think we should be working towards updating all our libraries and ditching deprecated libraries to create a more current and supported base state; the problem with the PicoContainer is that it is a central and complex component in our logic, which makes this process very difficult.)

[the remark this answers is not in the document]
Comment (srossbach) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: Saros.initComponent() just creates and attaches another container. It then does the IOC stuff and detaches the container. I really do not know if it is able to inject dependencies that are part of the session container. This method is not intended to be called over and over again(e.g 1000 times per second). It is more designed to offer access to the context for objects that are not instanciated by us, e.g in Saros/E we have several classes that are created by Eclipse at a specific time which is far behind the creation/activating and intializing the plugin context for our plugin.


As for the whole framework. We do not really use much stuff that PicoContainer offers. I already did take a look at Guice and came to the conclusion that a scoped context is much harder to implement with Guice. And as you already mentioned it would not be an easy task to switch the IOC framework and therefore I do not really see any benefits in doing it, but only wasting time.

[the remark this answers is not in the document]
Comment (tobous) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java]: How about making the variables and the injection static? The used objects should not change during the lifetime of the plugin.

```Java
@Inject
private static ProjectAPI projectAPI;

@Inject
private static EditorManager editorManager;

@Inject
private static Project project;

static {
    SarosPluginContext.initComponent(new SelectedEditorState());
}
```

As a side note, this looks kind of confusing with annotation injection. But we don't have `getComponent(Class)` for the SarosPluginContext. I still think we could add it, as proposed in [3295](https://saros-build.imp.fu-berlin.de/gerrit/#/c/3295/). It would also remove the false positives where IntelliJ states that variables using annotation injection are never set.
